### PR TITLE
Add replication primary primary

### PR DESCRIPTION
**Pull request by OliPou**
This change included:
- replication primary primary
- task `assert` to check all variables

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -17,23 +17,19 @@ server_id =
...
 |----------|-------------|---------|
 | `mysql_replication_gtid` | Whether or not mysql is using gtid for the replication | `true` |
-| `mysql_replication_role` | Defines whether a host is a `master` or a `slave` | mandatory |
 | `mysql_replication_user` | Defines the replication user | mandatory |
 | `mysql_socket` | MySQL Unix domain socket used for connections | `{{ __mysql_socket }}` |
 | `mysql_replication_master` | IP address or FQDN of the master | mandatory (for all slave servers)|
 | `mysql_master_user_gtid` | The value passed to [`master_use_gtid`][master-use-gtid] | `slave_pos` |
+| `mysql_master_user_gtid` | The value passed to [`master_use_gtid`][master-use-gtid] | `slave_pos` |
+| `mysql_replication_role_primary` | Define if the mysql server is primary | `false` |
+| `mysql_replication_role_secondary` | Define if the mysql server is primary | `false` |
```
Comment (twillouer) [inline README.md:31]: is secondary

Comment (OliPou) [reply README.md]: I just changed  `mandatory (for all slave servers)` to `mandatory (for all replica servers)`

File: README.md [hunk 2]
```diff
@@ -17,23 +17,18 @@ server_id =
...
 
 | Variable | Description | Default |
 |----------|-------------|---------|
 | `mysql_replication_gtid` | Whether or not mysql is using gtid for the replication | `true` |
-| `mysql_replication_role` | Defines whether a host is a `master` or a `slave` | mandatory |
 | `mysql_replication_user` | Defines the replication user | mandatory |
 | `mysql_socket` | MySQL Unix domain socket used for connections | `{{ __mysql_socket }}` |
-| `mysql_replication_master` | IP address or FQDN of the master | mandatory (for all slave servers)|
-| `mysql_master_user_gtid` | The value passed to [`master_use_gtid`][master-use-gtid] | `slave_pos` |
+| `mysql_replication_primary` | IP address or FQDN of the master | mandatory (for all replica servers)|
```
Comment (tleguern) [inline README.md:27]:
```suggestion
| `mysql_replication_primary` | IP address or FQDN of the primary | mandatory (for all replica servers)|
```

File: README.md [hunk 3]
```diff
@@ -51,59 +46,111 @@ Example:
...
 ```
 
 ## Dependencies
 
 None.
 
-## Example Playbook
+## Example Variable & Playbook
+
+### Replication Primary-Primary
```
Comment (tleguern) [inline README.md:59]: Isn't this better described as “multi primary”? https://dev.mysql.com/doc/refman/8.0/en/group-replication-multi-primary-mode.html

File: README.md [hunk 4]
```diff
@@ -51,59 +46,111 @@ Example:
...
+      expire_logs_days = 3
+      gtid-domain-id = 1
+      log_bin = binlog-m1
+mysql_replication_user:
+  name: replication
+  host: 192.68.25.%
+  password: "P@ssW0rD!"
+mysql_replication_primary: db2.exemple.com
+mysql_replication_role_primary: true
+
```
Comment (tleguern) [inline README.md:127]:
```suggestion
```

File: meta/main.yml [hunk 5]
```diff
@@ -5,7 +5,7 @@ galaxy_info:
   description: "MySQL role derived from Jeff Geerling for the master-slave replication"
   company: Deveryware
   license: "license (ISC)"
```
Comment (tleguern) [inline meta/main.yml:7]:
```suggestion
  license: "ISC"
```

File: meta/main.yml [hunk 6]
```diff
@@ -5,7 +5,7 @@ galaxy_info:
   description: "MySQL role derived from Jeff Geerling for the master-slave replication"
```
Comment (tleguern) [inline meta/main.yml:5]:
```suggestion
  description: "MySQL role implementing replication strategies"
```

File: tasks/assert.yml [hunk 7]
```diff
@@ -0,0 +1,92 @@
+---
+- name: test if rsyslog_receiver is set correctly
```
Comment (tleguern) [inline tasks/assert.yml:2]: Not quite.

File: tasks/assert.yml [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    - mysql_replication_user.priv is defined
+
+- name: test if mysql_replication_user.priv is set correctly
+  ansible.builtin.assert:
+    that:
+      - mysql_replication_user.priv is defined
+      - mysql_replication_user.priv is string
+    quiet: yes
+  when:
+    - mysql_replication_user.priv is defined
```
Comment (tleguern) [inline tasks/assert.yml:81]: Doing the assert dance with `mysql_replication_user` members make them mandatory, which contradict the README and the default values provided in `tasks/main.yml`.

File: tasks/assert.yml [hunk 9]
```diff
@@ -0,0 +1,92 @@
+---
+- name: test if rsyslog_receiver is set correctly
+  ansible.builtin.assert:
+    that:
+      - mysql_replication_gtid is defined
+      - mysql_replication_gtid is boolean
+    quiet: yes
+  when:
+    - mysql_replication_gtid is defined
```
Comment (tleguern) [inline tasks/assert.yml:9]: This is not needed on every variables, only the non mandatory without default value or the complex ones. Take another look at Robert de Bock's work and you will see the distinction.

Here the variable `mysql_replication_gtid` has a default value and is this always defined.

File: tasks/main.yml [hunk 10]
```diff
@@ -5,111 +5,127 @@
...
+
+- name: Ensure replication user exists on primary.
   mysql_user:
     name: "{{ mysql_replication_user.name }}"
     host: "{{ mysql_replication_user.host | default('%') }}"
     password: "{{ mysql_replication_user.password }}"
     priv: "{{ mysql_replication_user.priv | default('*.*:REPLICATION SLAVE,REPLICATION CLIENT') }}"
     state: present
     login_unix_socket: "{{ mysql_socket }}"
+  #no_log: true
```
Comment (tleguern) [inline tasks/main.yml:21]: You forgot the debug comment ;)

File: meta/main.yml (hunk 5 above)
Comment (tleguern) [reply meta/main.yml:7]: Or `ISC/MIT` if your code really contains bits from Jeff Geerling.

**Review by tleguern — CHANGES_REQUESTED** [review]
There are some bits that need more cleaning but it's good.

File: README.md (hunk 3 above)
Comment (OliPou) [reply README.md]: I didn't think of it! Done on the last version thanks :+1:

File: tasks/assert.yml (hunk 7 above)
Comment (OliPou) [reply tasks/assert.yml]: Delete in the last version.

File: tasks/assert.yml (hunk 9 above)
Comment (OliPou) [reply tasks/assert.yml]: Thanks for the information, I just removed the when.

File: tasks/main.yml (hunk 10 above)
Comment (OliPou) [reply tasks/main.yml]: Oh my bad! Corrected on the last version.

File: tasks/assert.yml (hunk 8 above)
Comment (OliPou) [reply tasks/assert.yml:71]: It should be alright on the last version.

File: README.md [hunk 11]
```diff
@@ -17,23 +17,18 @@ server_id =
...
-| `mysql_replication_role` | Defines whether a host is a `master` or a `slave` | mandatory |
 | `mysql_replication_user` | Defines the replication user | mandatory |
 | `mysql_socket` | MySQL Unix domain socket used for connections | `{{ __mysql_socket }}` |
-| `mysql_replication_master` | IP address or FQDN of the master | mandatory (for all slave servers)|
-| `mysql_master_user_gtid` | The value passed to [`master_use_gtid`][master-use-gtid] | `slave_pos` |
+| `mysql_replication_primary` | IP address or FQDN of the primary | mandatory (for all replica servers)|
+| `mysql_primary_use_gtid` | The value passed to [`primary_use_gtid`] | `slave_pos` |
+| `mysql_replication_role_primary` | Define if the server is the primary | `false` |
+| `mysql_replication_role_replica` | Define if the server is a replica | `false` |
+
```
Comment (tleguern) [inline README.md:31]: One extra blank to kill.

```suggestion
```

File: README.md [hunk 12]
```diff
@@ -51,59 +46,84 @@ Example:
...
+  host: 192.168.25.%
   password: "{{ vault_replication_password }}"
 ```
 
 ## Dependencies
 
 None.
 
-## Example Playbook
+## Example variable & playbook multi primary
```
Comment (tleguern) [inline README.md:57]: I would go even further and rename it for good:

```suggestion
## Example of a multi primary configuration
```

File: README.md [hunk 13]
```diff
@@ -51,59 +46,84 @@ Example:
...
+
+Playbook:
+
+```yaml
+- hosts: DBS
   roles:
     - role: ansible-role-mysql
-    - role: ansible-role-mysql-replication-master-slave
-```
+    - role: ansible-role-mysql-replication-primary-primary
```
Comment (tleguern) [inline README.md:124]:
```suggestion
    - role: ansible-role-mysql-replication
```

File: tasks/assert.yml [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+    quiet: yes
+
+- name: test if mysql_replication_user.host is set correctly
+  ansible.builtin.assert:
+    that:
+      - mysql_replication_user.host is defined
+      - mysql_replication_user.host is string
+    quiet: yes
+  when:
+    - mysql_replication_user.name is defined
```
Comment (tleguern) [inline tasks/assert.yml:46]: Should be `host`.

File: tasks/assert.yml [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+    - mysql_replication_user.priv is defined
+
+- name: test if mysql_replication_user.priv is set correctly
+  ansible.builtin.assert:
+    that:
+      - mysql_replication_user.priv is defined
+      - mysql_replication_user.priv is string
+    quiet: yes
+  when:
+    - mysql_replication_user.priv is defined
```
Comment (tleguern) [inline tasks/assert.yml:71]: This block is duplicated.

File: tasks/assert.yml [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+
+- name: test if mysql_replication_primary is set correctly
+  ansible.builtin.assert:
+    that:
+      - mysql_replication_primary is defined
+      - mysql_replication_primary is string
+    quiet: yes
+
+
```
Comment (tleguern) [inline tasks/assert.yml:80]: Can you kill the extra lines please?

**Review by tleguern — CHANGES_REQUESTED** [review]

File: tasks/assert.yml (hunk 7 above)
Comment (OliPou) [reply tasks/assert.yml]: That's should be done on the last version.

File: tasks/assert.yml (hunk 14 above)
Comment (OliPou) [reply tasks/assert.yml]: That's should be done on the last version

File: tasks/assert.yml (hunk 15 above)
Comment (OliPou) [reply tasks/assert.yml]: I just deleted that on the last version

File: tasks/assert.yml (hunk 16 above)
Comment (OliPou) [reply tasks/assert.yml]: That's should be done on the last version.

File: meta/main.yml [hunk 17]
```diff
@@ -2,10 +2,10 @@
 galaxy_info:
   author: Olivier Pouilly
   role_name: mysql-replication-master-slave
```
Comment (tleguern) [inline meta/main.yml:4]: Didn't catch this one on the first two readings. Dash is not a valid character in a role name for Galaxy.

```suggestion
  role_name: mysql_replication
```

**Review by tleguern — CHANGES_REQUESTED** [review]
Last bit.

**Review by tleguern — APPROVED** [review]
Looks good!

**Comment by tleguern** [comment]
Ok to merge @twillouer.

**Pull request merged**